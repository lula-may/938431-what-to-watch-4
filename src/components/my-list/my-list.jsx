import React from "react";
import {connect} from "react-redux";
import {Link} from "react-router-dom";
import PropTypes from "prop-types";

import MoviesList from "../movies-list/movies-list.jsx";

import {AppRoute} from "../../const.js";
import {ActionCreator, Operation} from "../../reducer/data/data.js";
import {getFavoriteMovies} from "../../reducer/data/selectors.js";
import {movieShape} from "../shapes.js";
import {getAvatarUrl} from "../../reducer/user/selectors.js";

const MyList = (props) => {
  const {avatarUrl, favoriteMovies, onMovieCardClick} = props;
  return (
    <div className="user-page">
      <header className="page-header user-page__head">
        <div className="logo">
          <Link to={AppRoute.ROOT} href="main.html" className="logo__link">
            <span className="logo__letter logo__letter--1">W</span>
            <span className="logo__letter logo__letter--2">T</span>
            <span className="logo__letter logo__letter--3">W</span>
          </Link>
        </div>

        <h1 className="page-title user-page__title">My list</h1>

        <div className="user-block">
          <div className="user-block__avatar">
            <img src={avatarUrl} alt="User avatar" width="63" height="63" />
          </div>
        </div>
      </header>

      <section className="catalog">
        <h2 className="catalog__title visually-hidden">Catalog</h2>
        <MoviesList
          movies={favoriteMovies}
          moviesCount={favoriteMovies.length}
          onMovieCardClick={onMovieCardClick}
        />
      </section>

      <footer className="page-footer">
        <div className="logo">
          <Link to={AppRoute.ROOT} href="main.html" className="logo__link logo__link--light">
            <span className="logo__letter logo__letter--1">W</span>
            <span className="logo__letter logo__letter--2">T</span>
            <span className="logo__letter logo__letter--3">W</span>
          </Link>
        </div>

        <div className="copyright">
          <p>Â© 2019 What to watch Ltd.</p>
        </div>
      </footer>
    </div>
  );
};

MyList.propTypes = {
  avatarUrl: PropTypes.string.isRequired,
  favoriteMovies: PropTypes.arrayOf(PropTypes.shape(movieShape)).isRequired,
  onMovieCardClick: PropTypes.func.isRequired,
};

const mapStateToProps = (state) => ({
  avatarUrl: getAvatarUrl(state),
  favoriteMovies: getFavoriteMovies(state),
});

const mapDispatchToProps = (dispatch) => ({
  onMovieCardClick(movie) {
    dispatch(ActionCreator.setActiveMovie(movie));
    dispatch(Operation.loadComments(movie));
  }
});

export {MyList};

export default connect(mapStateToProps, mapDispatchToProps)(MyList);
