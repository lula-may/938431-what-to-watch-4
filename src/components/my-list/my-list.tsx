import * as React from "react";
import {connect} from "react-redux";
import {Link} from "react-router-dom";
import PropTypes from "prop-types";

import MoviesList from "../movies-list/movies-list";

import {ActionCreator as StateActionCreator} from "../../reducer/app-state/app-state";
import {AppRoute} from "../../const";
import {getAvatarUrl} from "../../reducer/user/selectors";
import {getFavoriteMovies, getFavoriteLoadingError, getFavoriteLoadingState} from "../../reducer/data/selectors";
import {movieShape} from "../shapes";
import {Operation as DataOperation} from "../../reducer/data/data";

const messageStyle = {
  textAlign: `center`
};

class MyList extends React.PureComponent {

  componentDidMount() {
    const {loadFavoriteMovies} = this.props;
    loadFavoriteMovies();
  }

  render() {
    const {avatarUrl, favoriteMovies, hasFavoriteLoadingError, isLoading, onMovieCardClick} = this.props;

    return (
      <div className="user-page">
        <header className="page-header user-page__head">
          <div className="logo">
            <Link to={AppRoute.ROOT} href="main.html" className="logo__link">
              <span className="logo__letter logo__letter--1">W</span>
              <span className="logo__letter logo__letter--2">T</span>
              <span className="logo__letter logo__letter--3">W</span>
            </Link>
          </div>

          <h1 className="page-title user-page__title">My list</h1>

          <div className="user-block">
            <div className="user-block__avatar">
              <img src={avatarUrl} alt="User avatar" width="63" height="63" />
            </div>
          </div>
        </header>

        <section className="catalog">
          <h2 className="catalog__title visually-hidden">Catalog</h2>
          {isLoading && <p style={messageStyle}>Loading...</p>}
          {hasFavoriteLoadingError &&
            <p style={messageStyle}>Sorry, we failed to load your favortite movies. Please, try again later.</p>
          }
          <MoviesList
            movies={favoriteMovies}
            moviesCount={favoriteMovies.length}
            onMovieCardClick={onMovieCardClick}
          />
        </section>

        <footer className="page-footer">
          <div className="logo">
            <Link to={AppRoute.ROOT} href="main.html" className="logo__link logo__link--light">
              <span className="logo__letter logo__letter--1">W</span>
              <span className="logo__letter logo__letter--2">T</span>
              <span className="logo__letter logo__letter--3">W</span>
            </Link>
          </div>

          <div className="copyright">
            <p>Â© 2019 What to watch Ltd.</p>
          </div>
        </footer>
      </div>
    );
  }
}

MyList.propTypes = {
  avatarUrl: PropTypes.string.isRequired,
  favoriteMovies: PropTypes.arrayOf(PropTypes.shape(movieShape)).isRequired,
  hasFavoriteLoadingError: PropTypes.bool.isRequired,
  isLoading: PropTypes.bool.isRequired,
  loadFavoriteMovies: PropTypes.func.isRequired,
  onMovieCardClick: PropTypes.func.isRequired,
};

const mapStateToProps = (state) => ({
  avatarUrl: getAvatarUrl(state),
  favoriteMovies: getFavoriteMovies(state),
  hasFavoriteLoadingError: getFavoriteLoadingError(state),
  isLoading: getFavoriteLoadingState(state),
});

const mapDispatchToProps = (dispatch) => ({
  loadFavoriteMovies() {
    dispatch(DataOperation.loadFavoriteMovies());
  },
  onMovieCardClick(movie) {
    dispatch(StateActionCreator.setActiveMovie(movie));
    dispatch(DataOperation.loadComments(movie.id));
  }
});

export {MyList};

export default connect(mapStateToProps, mapDispatchToProps)(MyList);
