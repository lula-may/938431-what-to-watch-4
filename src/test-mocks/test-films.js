const movieTitles = [
  `Gone With The Wind`,
  `The Godfather`,
  `Casablanca`,
  `A Streetcar Named Desire`,
  `Some Likes It Hot`,
];

const movies = movieTitles.map((title, i) => {
  return {
    title,
    id: i,
    src: `poster-${i}.jpg`
  };
});

const headerMovie = {
  title: `Jurassic Park`,
  genre: `Sci-Fi`,
  releaseYear: 1993,
  bg: `background.jpg`,
  poster: `header-poster.jpg`
};

export {movies, headerMovie};

