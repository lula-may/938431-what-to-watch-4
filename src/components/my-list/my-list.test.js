import React from "react";
import {Router} from "react-router-dom";
import renderer from "react-test-renderer";

import {MyList} from "./my-list.jsx";
import history from "../../history.js";
import {testMovies} from "../../test-mocks/test-films.js";

const movies = testMovies.slice(0, 3);

describe(`MyList Component`, () => {
  it(`should render correctly MyList component`, () => {
    const tree = renderer.create(
        <Router history={history}>
          <MyList
            avatarUrl={`avatar.jpg`}
            favoriteMovies={movies}
            hasFavoriteLoadingError={false}
            isLoading={false}
            loadFavoriteMovies={() => {}}
            onMovieCardClick={() => {}}
          />
        </Router>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`should render Loading message`, () => {
    const tree = renderer.create(
        <Router history={history}>
          <MyList
            avatarUrl={`avatar.jpg`}
            favoriteMovies={[]}
            hasFavoriteLoadingError={false}
            isLoading={true}
            loadFavoriteMovies={() => {}}
            onMovieCardClick={() => {}}
          />
        </Router>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`should render Error message`, () => {
    const tree = renderer.create(
        <Router history={history}>
          <MyList
            avatarUrl={`avatar.jpg`}
            favoriteMovies={[]}
            hasFavoriteLoadingError={true}
            isLoading={false}
            loadFavoriteMovies={() => {}}
            onMovieCardClick={() => {}}
          />
        </Router>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

});
