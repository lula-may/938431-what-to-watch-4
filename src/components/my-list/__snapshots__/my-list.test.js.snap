// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyList Component should render Error message 1`] = `
<div
  className="user-page"
>
  <header
    className="page-header user-page__head"
  >
    <div
      className="logo"
    >
      <a
        className="logo__link"
        href="/"
        onClick={[Function]}
      >
        <span
          className="logo__letter logo__letter--1"
        >
          W
        </span>
        <span
          className="logo__letter logo__letter--2"
        >
          T
        </span>
        <span
          className="logo__letter logo__letter--3"
        >
          W
        </span>
      </a>
    </div>
    <h1
      className="page-title user-page__title"
    >
      My list
    </h1>
    <div
      className="user-block"
    >
      <div
        className="user-block__avatar"
      >
        <img
          alt="User avatar"
          height="63"
          src="avatar.jpg"
          width="63"
        />
      </div>
    </div>
  </header>
  <section
    className="catalog"
  >
    <h2
      className="catalog__title visually-hidden"
    >
      Catalog
    </h2>
    <p
      style={
        Object {
          "textAlign": "center",
        }
      }
    >
      Sorry, we failed to load your favortite movies. Please, try again later.
    </p>
    <div
      className="catalog__movies-list"
    />
  </section>
  <footer
    className="page-footer"
  >
    <div
      className="logo"
    >
      <a
        className="logo__link logo__link--light"
        href="/"
        onClick={[Function]}
      >
        <span
          className="logo__letter logo__letter--1"
        >
          W
        </span>
        <span
          className="logo__letter logo__letter--2"
        >
          T
        </span>
        <span
          className="logo__letter logo__letter--3"
        >
          W
        </span>
      </a>
    </div>
    <div
      className="copyright"
    >
      <p>
        © 2019 What to watch Ltd.
      </p>
    </div>
  </footer>
</div>
`;

exports[`MyList Component should render Loading message 1`] = `
<div
  className="user-page"
>
  <header
    className="page-header user-page__head"
  >
    <div
      className="logo"
    >
      <a
        className="logo__link"
        href="/"
        onClick={[Function]}
      >
        <span
          className="logo__letter logo__letter--1"
        >
          W
        </span>
        <span
          className="logo__letter logo__letter--2"
        >
          T
        </span>
        <span
          className="logo__letter logo__letter--3"
        >
          W
        </span>
      </a>
    </div>
    <h1
      className="page-title user-page__title"
    >
      My list
    </h1>
    <div
      className="user-block"
    >
      <div
        className="user-block__avatar"
      >
        <img
          alt="User avatar"
          height="63"
          src="avatar.jpg"
          width="63"
        />
      </div>
    </div>
  </header>
  <section
    className="catalog"
  >
    <h2
      className="catalog__title visually-hidden"
    >
      Catalog
    </h2>
    <p
      style={
        Object {
          "textAlign": "center",
        }
      }
    >
      Loading...
    </p>
    <div
      className="catalog__movies-list"
    />
  </section>
  <footer
    className="page-footer"
  >
    <div
      className="logo"
    >
      <a
        className="logo__link logo__link--light"
        href="/"
        onClick={[Function]}
      >
        <span
          className="logo__letter logo__letter--1"
        >
          W
        </span>
        <span
          className="logo__letter logo__letter--2"
        >
          T
        </span>
        <span
          className="logo__letter logo__letter--3"
        >
          W
        </span>
      </a>
    </div>
    <div
      className="copyright"
    >
      <p>
        © 2019 What to watch Ltd.
      </p>
    </div>
  </footer>
</div>
`;

exports[`MyList Component should render correctly MyList component 1`] = `
<div
  className="user-page"
>
  <header
    className="page-header user-page__head"
  >
    <div
      className="logo"
    >
      <a
        className="logo__link"
        href="/"
        onClick={[Function]}
      >
        <span
          className="logo__letter logo__letter--1"
        >
          W
        </span>
        <span
          className="logo__letter logo__letter--2"
        >
          T
        </span>
        <span
          className="logo__letter logo__letter--3"
        >
          W
        </span>
      </a>
    </div>
    <h1
      className="page-title user-page__title"
    >
      My list
    </h1>
    <div
      className="user-block"
    >
      <div
        className="user-block__avatar"
      >
        <img
          alt="User avatar"
          height="63"
          src="avatar.jpg"
          width="63"
        />
      </div>
    </div>
  </header>
  <section
    className="catalog"
  >
    <h2
      className="catalog__title visually-hidden"
    >
      Catalog
    </h2>
    <div
      className="catalog__movies-list"
    >
      <article
        className="small-movie-card catalog__movies-card"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <a
          className="small-movie-card__image"
          href="/films/0"
          onClick={[Function]}
        >
          <video
            height="175"
            poster="poster-0.jpg"
            width="280"
          />
        </a>
        <h3
          className="small-movie-card__title"
        >
          <a
            className="small-movie-card__link"
            href="/films/0"
            onClick={[Function]}
          >
            Jurassic Park
          </a>
        </h3>
      </article>
      <article
        className="small-movie-card catalog__movies-card"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <a
          className="small-movie-card__image"
          href="/films/1"
          onClick={[Function]}
        >
          <video
            height="175"
            poster="poster-1.jpg"
            width="280"
          />
        </a>
        <h3
          className="small-movie-card__title"
        >
          <a
            className="small-movie-card__link"
            href="/films/1"
            onClick={[Function]}
          >
            Gone With The Wind
          </a>
        </h3>
      </article>
      <article
        className="small-movie-card catalog__movies-card"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <a
          className="small-movie-card__image"
          href="/films/2"
          onClick={[Function]}
        >
          <video
            height="175"
            poster="poster-2.jpg"
            width="280"
          />
        </a>
        <h3
          className="small-movie-card__title"
        >
          <a
            className="small-movie-card__link"
            href="/films/2"
            onClick={[Function]}
          >
            The Godfather
          </a>
        </h3>
      </article>
    </div>
  </section>
  <footer
    className="page-footer"
  >
    <div
      className="logo"
    >
      <a
        className="logo__link logo__link--light"
        href="/"
        onClick={[Function]}
      >
        <span
          className="logo__letter logo__letter--1"
        >
          W
        </span>
        <span
          className="logo__letter logo__letter--2"
        >
          T
        </span>
        <span
          className="logo__letter logo__letter--3"
        >
          W
        </span>
      </a>
    </div>
    <div
      className="copyright"
    >
      <p>
        © 2019 What to watch Ltd.
      </p>
    </div>
  </footer>
</div>
`;
